### PR TITLE
[SYCLomatic] Support migration for  cub::{StoreDirectBlocked, StoreDirectStriped} API

### PR DESCRIPTION
**Pull request by abhilash1910**
Add Migration of cub::store  API . Linked to #1819   
cc @yihanwg  , @zhimingwang36

### CODE REVIEW

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 1]
```diff
@@ -1170,7 +1170,7 @@ void ExprAnalysis::analyzeType(TypeLoc TL, const Expr *CSCE,
       }
     }
     if (OS.str() != "cub::WarpScan" && OS.str() != "cub::WarpReduce" &&
-        OS.str() != "cub::BlockReduce" && OS.str() != "cub::BlockScan") {
+        OS.str() != "cub::BlockReduce" && OS.str() != "cub::BlockScan" && OS.str != "cub::BlockLoad") {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ExprAnalysis.cpp]:
```suggestion
        OS.str() != "cub::BlockReduce" && OS.str() != "cub::BlockScan" && OS.str() != "cub::BlockLoad") {
```

File: clang/lib/DPCT/APINames_CUB.inc [hunk 2]
```diff
@@ -123,6 +122,7 @@ ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Offset, Offset, fals
 ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Rotate, Rotate, false, NO_FLAG, P4, "Comment")
 ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Up, Up, false, NO_FLAG, P4, "Comment")
 ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Down, Down, false, NO_FLAG, P4, "Comment")
+ENTRY_MEMBER_FUNCTION(cub::BlockLoad, cub::BlockLoad, Load, Load, true, NO_FLAG, P4, "Comment")
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APINames_CUB.inc]: This file just used to check migration status.
```suggestion
ENTRY_MEMBER_FUNCTION(cub::BlockLoad, cub::BlockLoad, Load, Load, true, NO_FLAG, P4, "Successful")
```

File: clang/test/dpct/cub/blocklevel/blockload.cu [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    dim3 grid((N + BlockSize - 1) / BlockSize);
+
+    BlockLoadKernel<<<grid, block>>>(d_data);
+    cudaDeviceSynchronize();
+    //verify_data(d_data, N);   
+
+    cudaFree(d_data);
+
+    return 0;
+}
```
Comment (yihwang-nv) [inline clang/test/dpct/cub/blocklevel/blockload.cu]: Please add a new line at the end of this file

File: clang/test/dpct/cub/blocklevel/blockload.cu [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  return;
+}
+void print_data(int* data, int num) {
+  for (int i = 0; i < num; i++) {
+    std::cout << data[i] << " ";
+  }
+  std::cout << std::endl;
+}
+
+//CHECK: void SumKernel(int* data,
```
Comment (yihwang-nv) [inline clang/test/dpct/cub/blocklevel/blockload.cu]: This lit need update, usually, we check migration result here, and make sure the result is correct.

Comment (abhilash1910) [reply clang/test/dpct/cub/blocklevel/blockload.cu:33]: Do we need to add a sycl kernel for blockload here? for reduce /exclusive /inclusive scans we have sycl::reduce_over_group etc apis . Do we need such api specifically for load as well?

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 5]
```diff
@@ -1328,7 +1342,8 @@ void CubRule::processTypeLoc(const TypeLoc *TL) {
                                       MapNames::getClNamespace() + "sub_group",
                                       SM));
   } else if (TypeName.find("class cub::BlockScan") == 0 ||
-             TypeName.find("class cub::BlockReduce") == 0) {
+             TypeName.find("class cub::BlockReduce") == 0 ||
+             ObjTypeStr.find("class cub::BlockLoad") == 0) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/CUBAPIMigration.cpp]:
```suggestion
             TypeName.find("class cub::BlockLoad") == 0) {
```

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 6]
```diff
@@ -1139,6 +1141,17 @@ void CubRule::processBlockLevelMemberCall(const CXXMemberCallExpr *BlockMC) {
     CubParamAs << GroupOrWorkitem << InEA.getReplacedString() << OpRepl;
     Repl = NewFuncName + "(" + ParamList + ")";
     emplaceTransformation(new ReplaceStmt(BlockMC, Repl));
+  } else if (FuncName == "Load") {
+  
+    GroupOrWorkitem = DpctGlobalInfo::getItem(BlockMC);
+    NewFuncName = Mapnames::getClNamespace() + "load";
```
Comment (yihwang-nv) [inline clang/lib/DPCT/CUBAPIMigration.cpp]:
```suggestion
    NewFuncName = MapNames::getClNamespace() + "load";
```

Comment (yihwang-nv) [inline clang/lib/DPCT/CUBAPIMigration.cpp]: The mapping seems not correct, I try to compile lit int this PR, but it failed.

Comment (abhilash1910) [reply clang/lib/DPCT/CUBAPIMigration.cpp]: I added in the Load mapping in the expression (cxxmembercallexpr).

**Comment by abhilash1910** [comment]
Linked to #2094

File: clang/test/dpct/cub/intrinsic/store.cu [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  // CHECK: dpct::group::load_striped(item_ct1, d_data, thread_data);
+  cub::StoreDirectStriped<128>(threadIdx.x, d_data, thread_data);
+}
+
+
+__global__ void BlockedToStripedKernel(int *d_data) {
+  int thread_data[4];
+  // CHECK: dpct::group::load_blocked(item_ct1, d_data, thread_data);
+  cub::StoreDirectBlocked(threadIdx.x, d_data, thread_data);
+}
```
Comment (yihwang-nv) [inline clang/test/dpct/cub/intrinsic/store.cu]: Need a newline at the end of file.

Comment (abhilash1910) [reply clang/test/dpct/cub/intrinsic/store.cu]: Added, please help trigger CI

File: clang/lib/DPCT/APINames_CUB.inc [hunk 8]
```diff
@@ -107,7 +107,6 @@ ENTRY_MEMBER_FUNCTION(cub::BlockExchange, cub::BlockExchange, ScatterToStripedFl
 ENTRY_MEMBER_FUNCTION(cub::BlockHistogram, cub::BlockHistogram, InitHistogram, InitHistogram, false, NO_FLAG, P4, "Comment")
 ENTRY_MEMBER_FUNCTION(cub::BlockHistogram, cub::BlockHistogram, Histogram, Histogram, false, NO_FLAG, P4, "Comment")
 ENTRY_MEMBER_FUNCTION(cub::BlockHistogram, cub::BlockHistogram, Composite, Composite, false, NO_FLAG, P4, "Comment")
-ENTRY_MEMBER_FUNCTION(cub::BlockLoad, cub::BlockLoad, Load, Load, false, NO_FLAG, P4, "Comment")
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APINames_CUB.inc]: Why we remove this?

File: clang/lib/DPCT/APINames_CUB.inc [hunk 9]
```diff
@@ -124,7 +123,6 @@ ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Rotate, Rotate, fals
 ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Up, Up, false, NO_FLAG, P4, "Comment")
 ENTRY_MEMBER_FUNCTION(cub::BlockShuffle, cub::BlockShuffle, Down, Down, false, NO_FLAG, P4, "Comment")
 
-
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APINames_CUB.inc]: Let's remove extra blank line change.

File: clang/lib/DPCT/APINames_CUB.inc [hunk 10]
```diff
@@ -192,10 +190,10 @@ ENTRY(cub::ThreadLoad, cub::ThreadLoad, true, NO_FLAG, P4, "Successful")
 ENTRY(cub::ThreadStore, cub::ThreadStore, true, NO_FLAG, P4, "Successful")
 ENTRY(cub::LoadDirectBlocked, cub::LoadDirectBlocked, false, NO_FLAG, P4, "Comment")
 ENTRY(cub::LoadDirectBlockedVectorized, cub::LoadDirectBlockedVectorized, false, NO_FLAG, P4, "Comment")
-ENTRY(cub::StoreDirectBlocked, cub::StoreDirectBlocked, false, NO_FLAG, P4, "Comment")
+ENTRY(cub::StoreDirectBlocked, cub::StoreDirectBlocked, true, NO_FLAG, P4, "Comment")
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APINames_CUB.inc]: The comment must be one of "Successful" or "Partial".

**Review by yihwang-nv — CHANGES_REQUESTED** [review]

**Comment by yihwang-nv** [comment]
Can you merge this PR into https://github.com/oneapi-src/SYCLomatic/pull/1819?

File: clang/lib/DPCT/APINames_CUB.inc (hunk 10 above)
Comment (abhilash1910) [reply clang/lib/DPCT/APINames_CUB.inc]: Yes added

**Pull request closed without merging**